src/cc/usdt/usdt.cc

```diff
@@ -295,24 +295,46 @@ Probe *Context::get(const std::string &provider_name,
 
 bool Context::enable_probe(const std::string &probe_name,
                            const std::string &fn_name) {
+  return enable_probe("", probe_name, fn_name);
+}
+
+bool Context::enable_probe(const std::string &provider_name,
+                           const std::string &probe_name,
+                           const std::string &fn_name) {
   if (pid_stat_ && pid_stat_->is_stale())
     return false;
 
-  // FIXME: we may have issues here if the context has two same probes's
-  // but different providers. For example, libc:setjmp and rtld:setjmp,
-  // libc:lll_futex_wait and rtld:lll_futex_wait.
+  unsigned int matches = 0;
   Probe *found_probe = nullptr;
   for (auto &p : probes_) {
     if (p->name_ == probe_name) {
-      if (found_probe != nullptr) {
-         fprintf(stderr, "Two same-name probes (%s) but different providers\n",
-                 probe_name.c_str());
-         return false;
+      if (found_probe == nullptr && provider_name == "")
+      {
+        found_probe = p.get();
+        matches++;
+      }
+      else if (found_probe != nullptr && provider_name == "")
+      {
+        fprintf(stderr, "Found duplicate provider (%s) for underspecified probe (%s)\n",
```

Comment:
Why do we need these logic? I mean it's totally normal for Tracepoints from two different providers have the same name (which is exactly the problem we are solving).

Also these `if` conditions are quite twisted... How about:
```
if ((p->name_ == probe_name) && (provider_name.empty() || (provider_name == p->provider()))) {
  do_normal_found_probe_work;
  if (provider_name.empty()) {
    do_some_work_to_warn_people;
  }
}
```

Reply:
Hmm yes I agree it can be simplified. I thought it would be helpful to indicate where the duplicate providers are if they didn't specify the provider name, but we can just leave it up to the caller to figure out.

Reply:
I followed this approach of checking all of the conditions, and setting the pointer if it's unset, otherwise logging an error.

This is the behavior it does now, and the enhancement that I've added is that if a provider is specified it will use it and avoid the collision.